### DISCUSSION TITLE
FastAPI shows no attributes for models when a  @model_serializer is used.

### DISCUSSION THREAD

**Discussion by TNick** [fastapi/fastapi#10661] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Any, Dict
from fastapi import FastAPI
from pydantic import BaseModel, Field, SerializationInfo, model_serializer


class SomeModel(BaseModel):
    lorem: str = Field(..., description="describe me")
    ipsum: int

    @model_serializer(when_used="always", mode="wrap")
    def clean_model(self, serializer: Any, info: SerializationInfo) -> Dict[str, Any]:
        data: Dict[str, Any] = serializer(self)
        return data

app = FastAPI()

@app.get("/another/{dataset}", response_model=SomeModel)
def get_another_dataset(dataset: str):
    raise NotImplementedError
```

### Description

* run provided code
* go to /docs

![image](https://github.com/tiangolo/fastapi/assets/1712330/5cd68ca2-aa4f-434c-93a7-c4dcd35829b6)

### Operating System

Linux, Windows, macOS, Other

### FastAPI Version

0.104.1

### Pydantic Version

2.5.1

### Python Version

3.9

### Additional Context

Uncovered while investigating [this geojson-pydantic  issue](https://github.com/developmentseed/geojson-pydantic/issues/150).

Edit: code sandbox available [here](https://codesandbox.io/p/sandbox/dreamy-neumann-xqscpv).

**Comment by volbil** [comment]
I'm facing same issue. I think it can be narroved down to Pydantic's model with `@model_serializer` passed to FastAPI's `response_model` field. Models used for endpoint arguments do display attributes.

**Comment by DLukeNelson** [comment]
I ran into this recently, and I found a workaround. In my case, removing the `Dict[str, Any]` type hint for `clean_model`'s return value fixed the issue for me.
I couldn't say why this works, I just happened to be trying to create an MVE, and left off the type hint. Surprisingly, it worked correctly.

**Reply by alex-pobeditel-2004** [reply]
@DLukeNelson you're my savior!  
After move to Pydantic v2, all my models (that use common base model) lost OpenAPI response model schemas - now I see that happened because of serializer that I added because pydantic didn't serialize Path and AnyUrl objects correctly.

Without type annotation on serializer, everything works like a charm!

@tiangolo @Kludex I don't know if this is a FastAPI or Pydantic issue, but I hope that real problem solution will make it to release once :)

**Comment by eseglem** [comment, marked as answer]
The best workaround I have found is to not type hint the return on the serializer. Then it will use the same definition as it would without the serializer defined. 

```python
    def clean_model(self, serializer: Any, info: SerializationInfo):  # type: ignore [no-untyped-def]
```

From the FastAPI side of things this is because it calls `.model_json_schema(mode="serialization")` and typing the return of the serializer changes what pydantic returns for the serialization mode. I feel like what FastAPI is doing is correct and the root of the issue is on the pydantic side. I wrote up an issue on pydantic requesting a nicer way to handle this: https://github.com/pydantic/pydantic/issues/8791 if anyone wants thumbs up it or chime in with thoughts.
